src/store/slices/balancesSlice.ts

```diff
@@ -0,0 +1,193 @@
+import { AnchorProvider } from '@coral-xyz/anchor'
+import { PayloadAction, createAsyncThunk, createSlice } from '@reduxjs/toolkit'
+import { Cluster } from '@solana/web3.js'
+import { CSAccount } from '../../storage/cloudStorage'
+import { Balances } from '../../types/solana'
+import * as solUtils from '../../utils/solanaUtils'
+import { getBalanceHistory, getTokenPrices } from '../../utils/walletApiV2'
+import { Prices } from '../../types/balance'
+import { AccountBalance } from '../../generated/graphql'
+import { getEscrowTokenAccount } from '../../utils/solanaUtils'
+
+type BalancesByWallet = Record<string, Balances>
+type BalancesByCluster = Record<Cluster, BalancesByWallet>
+
+type BalanceHistoryByCurrency = Record<string, AccountBalance[]>
+type BalanceHistoryByWallet = Record<string, BalanceHistoryByCurrency>
+type BalanceHistoryByCluster = Record<Cluster, BalanceHistoryByWallet>
+
+export type BalancesState = {
+  tokens: BalancesByCluster
+  balancesLoading?: boolean
+  tokenPrices?: Prices
+  balanceHistory: BalanceHistoryByCluster
+}
+
+const initialBalances = {
+  sol: { balance: 0, tokenAccount: '' },
+  mobile: { balance: 0, tokenAccount: '' },
+  dc: { balance: 0, tokenAccount: '' },
+  iot: { balance: 0, tokenAccount: '' },
+  hnt: { balance: 0, tokenAccount: '' },
+  dcEscrow: { balance: 0, tokenAccount: '' },
+}
+
+const initialState: BalancesState = {
+  tokens: {
+    'mainnet-beta': {},
+    devnet: {},
+    testnet: {},
+  },
+  balanceHistory: {
+    'mainnet-beta': {},
+    devnet: {},
+    testnet: {},
+  },
+}
+
+export const readTokenBalances = createAsyncThunk(
```

Comment:
Only thing that worries me here is how this will translate when we add other SPL tokens to the app 🤔